Core/include/Acts/TrackFitting/MBFSmoother.hpp

```diff
@@ -0,0 +1,111 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2024 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/TrackParametrization.hpp"
+#include "Acts/EventData/MultiTrajectory.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/TrackFitting/KalmanFitterError.hpp"
+#include "Acts/Utilities/Delegate.hpp"
+#include "Acts/Utilities/Logger.hpp"
+#include "Acts/Utilities/Result.hpp"
+
+#include <cassert>
+#include <cstddef>
+#include <system_error>
+
+namespace Acts {
+
+/// Kalman trajectory smoother based on the Modified Bryson–Frazier smoother.
+///
+/// This implements not a single smoothing step, but the full backwards
+/// smoothing procedure for a filtered, forward trajectory using the stored
+/// linearization.
+class MBFSmoother {
+ public:
+  /// Run the Kalman smoothing for one trajectory.
+  ///
+  /// @param[in,out] trajectory The trajectory to be smoothed
+  /// @param[in] entryIndex The index of state to start the smoothing
+  /// @param[in] logger Where to write logging information to
+  template <typename traj_t>
+  Result<void> operator()(const GeometryContext& /*gctx*/, traj_t& trajectory,
+                          std::size_t entryIndex,
+                          const Logger& /*logger*/ = getDummyLogger()) const {
+    using TrackStateProxy = typename traj_t::TrackStateProxy;
+
+    TrackStateProxy start_ts = trajectory.getTrackState(entryIndex);
+
+    BoundMatrix big_lambda_hat = BoundMatrix::Zero();
+    BoundVector small_lambda_hat = BoundVector::Zero();
+
+    trajectory.applyBackwards(start_ts.index(), [&](TrackStateProxy ts) {
```

Comment:
This will be heavy. If we want to merge this, it should go into a `.cpp` file using some mechanism like I used for the GainMatrixSmoother.

Reply:
Right this I will definitely change